File: Firebase/iOS/FirebaseAuth.swift
import FirebaseAuth

/// Apple-native Firebase Authentication adapter for iOS and macOS.
public final class PawsomeFirebaseAuth {
    public static let shared = PawsomeFirebaseAuth()

    private let auth = Auth.auth()

    private init() {}

    public var currentUser: User? { auth.currentUser }
    public var currentUid: String? { auth.currentUser?.uid }

    public func signInWithGoogle(idToken: String, accessToken: String? = nil) async throws -> User {
        let credential = GoogleAuthProvider.credential(
            withIDToken: idToken,
            accessToken: accessToken
        )
        let result = try await auth.signIn(with: credential)
        return result.user
    }

    public func signInWithTwitter(
        accessToken: String,
        secret: String
    ) async throws -> User {
        let credential = OAuthProvider.credential(
            withProviderID: "twitter.com",
            accessToken: accessToken,
            secret: secret
        )
        let result = try await auth.signIn(with: credential)
        return result.user
    }

    public func getIDToken(forceRefresh: Bool = false) async throws -> String {
        let user = auth.currentUser
            ?? NSError(domain: "PawsomeFirebaseAuth", code: 1,
                       userInfo: [NSLocalizedDescriptionKey: "Not signed in"])
        return try await user.getIDToken(forcingRefresh: forceRefresh)
    }

    public func signOut() throws {
        try auth.signOut()
    }
}

File: Firebase/Android/FirebaseAuth.kt
package com.example.pawsome.firebase

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.OAuthProvider
import kotlinx.coroutines.tasks.await

/**
 * Android-native Firebase Authentication adapter.
 *
 * GoogleAuth.kt remains responsible for obtaining the Google ID token. This
 * adapter hands that token to the Firebase Android Authentication SDK.
 */
class FirebaseAuth {
    private val auth = FirebaseAuth.getInstance()

    val currentUser get() = auth.currentUser
    val currentUid: String? get() = auth.currentUser?.uid

    suspend fun signInWithGoogle(googleIdToken: String) =
        auth.signInWithCredential(
            GoogleAuthProvider.getCredential(googleIdToken, null)
        ).await().user
            ?: error("Firebase returned no user after Google sign-in.")

    suspend fun signInWithTwitter(oauthToken: String, oauthTokenSecret: String) =
        auth.signInWithCredential(
            OAuthProvider.newCredentialBuilder("twitter.com")
                .setAccessToken(oauthToken)
                .setSecret(oauthTokenSecret)
                .build()
        ).await().user
            ?: error("Firebase returned no user after Twitter sign-in.")

    suspend fun refreshToken(): String {
        val user = auth.currentUser ?: error("Not signed in")
        return user.getIdToken(true).await()?.token
            ?: error("Firebase returned no ID token.")
    }

    fun signOut() = auth.signOut()
}
